File: sdk/go/client_test.go
package farfield

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"net/http/httptest"
	"reflect"
	"sync"
	"testing"
	"time"
)

func TestCaptureRetriesExactBodyAndAppliesScope(t *testing.T) {
	t.Parallel()
	var mu sync.Mutex
	var bodies [][]byte
	server := httptest.NewServer(http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
		if request.URL.Path != "/v1/history/records" || request.Header.Get("Authorization") != "Bearer secret" || request.Header.Get("User-Agent") != "farfield-go/"+Version {
			t.Errorf("request = %s, headers = %#v", request.URL.Path, request.Header)
		}
		var body json.RawMessage
		if err := json.NewDecoder(request.Body).Decode(&body); err != nil {
			t.Error(err)
		}
		mu.Lock()
		bodies = append(bodies, append([]byte(nil), body...))
		attempt := len(bodies)
		mu.Unlock()
		writer.Header().Set("Content-Type", "application/json")
		if attempt == 1 {
			writer.WriteHeader(http.StatusServiceUnavailable)
			_, _ = writer.Write([]byte(`{"error":{"code":"FH_UNAVAILABLE","message":"retry","retryable":true}}`))
			return
		}
		var value map[string]any
		_ = json.Unmarshal(body, &value)
		_ = json.NewEncoder(writer).Encode(map[string]any{"id": value["id"], "conversation_id": value["conversation_id"], "kind": value["kind"]})
	}))
	defer server.Close()

	client, err := New(
		WithEndpoint(server.URL), WithToken("secret"), WithRetries(1, 0),
		WithDefaults(Scope{Agent: "researcher", Tags: map[string]string{"env": "test", "shared": "default"}}),
		WithBeforeSend(func(_ context.Context, input CaptureInput) (*CaptureInput, error) {
			content := input.Content.(map[string]any)
			delete(content, "secret")
			input.Content = content
			return &input, nil
		}),
	)
	if err != nil {
		t.Fatal(err)
	}
	ctx := WithScope(context.Background(), Scope{ConversationID: "conv_go", TraceID: "trace_go", Tags: map[string]string{"shared": "scope"}})
	record, err := client.Capture(ctx, CaptureInput{
		Kind: "model.response", Content: map[string]any{"text": "hello", "secret": "remove"},
		Tags: map[string]string{"request": "42"},
	})
	if err != nil {
		t.Fatal(err)
	}
	if record.ID == "" || record.ConversationID != "conv_go" {
		t.Fatalf("record = %#v", record)
	}
	if len(bodies) != 2 || !reflect.DeepEqual(bodies[0], bodies[1]) {
		t.Fatalf("retry bodies differ: %q, %q", bodies[0], bodies[1])
	}
	var sent map[string]any
	if err := json.Unmarshal(bodies[0], &sent); err != nil {
		t.Fatal(err)
	}
	if sent["agent"] != "researcher" || sent["trace_id"] != "trace_go" || sent["occurred_at"] == nil {
		t.Fatalf("sent = %#v", sent)
	}
	tags := sent["tags"].(map[string]any)
	if tags["env"] != "test" || tags["shared"] != "scope" || tags["request"] != "42" {
		t.Fatalf("tags = %#v", tags)
	}
	content := sent["content"].(map[string]any)
	if _, exists := content["secret"]; exists {
		t.Fatalf("before-send did not redact content: %#v", content)
	}
}

func TestCaptureBatchAndConversationHelpers(t *testing.T) {
	t.Parallel()
	server := httptest.NewServer(http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
		if request.URL.Path != "/v1/history/segments" {
			t.Errorf("path = %s", request.URL.Path)
		}
		var value struct {
			ID      string         `json:"id"`
			Records []CaptureInput `json:"records"`
		}
		if err := json.NewDecoder(request.Body).Decode(&value); err != nil {
			t.Error(err)
		}
		if value.ID == "" || len(value.Records) != 2 || value.Records[0].ID == "" || value.Records[0].ConversationID != "conv_batch" {
			t.Errorf("batch = %#v", value)
		}
		_ = json.NewEncoder(writer).Encode(map[string]any{"id": value.ID, "conversation_id": "conv_batch", "entries": []any{map[string]any{}, map[string]any{}}})
	}))
	defer server.Close()
	client, _ := New(WithEndpoint(server.URL), WithRetries(0, 0))
	segment, err := client.Conversation("conv_batch").CaptureBatch(context.Background(), []CaptureInput{
		{Kind: "message.input", Content: map[string]any{"text": "hello"}},
		{Kind: "message.output", Content: map[string]any{"text": "hi"}},
	})
	if err != nil {
		t.Fatal(err)
	}
	if segment.ID == "" || segment.ConversationID != "conv_batch" || len(segment.Entries) != 2 {
		t.Fatalf("segment = %#v", segment)
	}
}

func TestBeforeSendCanDrop(t *testing.T) {
	t.Parallel()
	client, err := New(WithBeforeSend(func(context.Context, CaptureInput) (*CaptureInput, error) { return nil, nil }))
	if err != nil {
		t.Fatal(err)
	}
	_, err = client.Capture(context.Background(), CaptureInput{ConversationID: "conv", Kind: "private", Content: "secret"})
	if !errors.Is(err, ErrDropped) {
		t.Fatalf("error = %v", err)
	}
}

func TestBackgroundProcessorBatchesByConversationAndFlushes(t *testing.T) {
	t.Parallel()
	var mu sync.Mutex
	var segments []BatchInput
	server := httptest.NewServer(http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
		var value BatchInput
		if err := json.NewDecoder(request.Body).Decode(&value); err != nil {
			t.Error(err)
		}
		mu.Lock()
		segments = append(segments, value)
		mu.Unlock()
		_ = json.NewEncoder(writer).Encode(map[string]any{
			"id": value.ID, "conversation_id": value.Records[0].ConversationID, "entries": []any{},
		})
	}))
	defer server.Close()
	client, _ := New(WithEndpoint(server.URL), WithRetries(0, 0), WithDefaults(Scope{Tags: map[string]string{"env": "test"}}))
	processor, err := NewBackgroundProcessor(client, ProcessorOptions{MaxBatchSize: 10, ScheduleDelay: time.Millisecond})
	if err != nil {
		t.Fatal(err)
	}
	ctx := WithScope(context.Background(), Scope{ConversationID: "conv_one", Agent: "researcher"})
	for _, kind := range []string{"message.user", "message.assistant"} {
		accepted, submitErr := processor.Submit(ctx, CaptureInput{Kind: kind, Content: kind})
		if submitErr != nil || !accepted {
			t.Fatalf("submit = %v, %v", accepted, submitErr)
		}
	}
	accepted, err := processor.Submit(context.Background(), CaptureInput{ConversationID: "conv_two", Kind: "tool.result", Tool: "search", Content: true})
	if err != nil || !accepted {
		t.Fatalf("submit = %v, %v", accepted, err)
	}
	flushCtx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()
	if err := processor.Flush(flushCtx); err != nil {
		t.Fatal(err)
	}
	if err := processor.Shutdown(flushCtx); err != nil {
		t.Fatal(err)
	}
	mu.Lock()
	defer mu.Unlock()
	if len(segments) != 2 {
		t.Fatalf("segments = %#v", segments)
	}
	counts := map[string]int{}
	for _, segment := range segments {
		counts[segment.Records[0].ConversationID] = len(segment.Records)
		if segment.Records[0].ConversationID == "conv_one" && (segment.Records[0].Agent != "researcher" || segment.Records[0].Tags["env"] != "test") {
			t.Fatalf("snapshotted record = %#v", segment.Records[0])
		}
	}
	if counts["conv_one"] != 2 || counts["conv_two"] != 1 {
		t.Fatalf("counts = %#v", counts)
	}
	stats := processor.Stats()
	if stats.Enqueued != 3 || stats.Committed != 3 || stats.Pending != 0 || stats.Batches != 2 {
		t.Fatalf("stats = %#v", stats)
	}
}

func TestBackgroundProcessorReportsDeliveryFailures(t *testing.T) {
	t.Parallel()
	server := httptest.NewServer(http.HandlerFunc(func(writer http.ResponseWriter, _ *http.Request) {
		writer.WriteHeader(http.StatusServiceUnavailable)
		_, _ = writer.Write([]byte(`{"error":{"code":"FH_BUSY","message":"busy","retryable":true}}`))
	}))
	defer server.Close()
	client, _ := New(WithEndpoint(server.URL), WithRetries(0, 0))
	errorsSeen := make(chan error, 1)
	processor, _ := NewBackgroundProcessor(client, ProcessorOptions{ScheduleDelay: time.Millisecond, OnError: func(err error) { errorsSeen <- err }})
	accepted, err := processor.Submit(context.Background(), CaptureInput{ConversationID: "conv_fail", Kind: "message.user", Content: "hello"})
	if err != nil || !accepted {
		t.Fatalf("submit = %v, %v", accepted, err)
	}
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()
	if err := processor.Flush(ctx); err == nil {
		t.Fatal("flush unexpectedly succeeded")
	}
	if err := processor.Shutdown(ctx); err == nil {
		t.Fatal("shutdown unexpectedly succeeded after a permanent delivery failure")
	}
	select {
	case <-errorsSeen:
	default:
		t.Fatal("on-error callback was not called")
	}
	if stats := processor.Stats(); stats.Failed != 1 || stats.Committed != 0 {
		t.Fatalf("stats = %#v", stats)
	}
}

func TestHistoryReadSurface(t *testing.T) {
	t.Parallel()
	server := httptest.NewServer(http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
		writer.Header().Set("Content-Type", "application/json")
		switch request.URL.Path {
		case "/v1/health":
			_, _ = writer.Write([]byte(`{"ok":true,"service":"farfield"}`))
		case "/v1/history/records":
			if request.URL.Query().Get("conversation_id") != "conv_read" || request.URL.Query().Get("limit") != "25" || request.URL.Query().Get("tag") != "env=test" {
				t.Errorf("query = %s", request.URL.RawQuery)
			}
			_, _ = writer.Write([]byte(`[{"id":"rec_read","conversation_id":"conv_read","kind":"message.user"}]`))
		case "/v1/history/records/rec_read":
			_, _ = writer.Write([]byte(`{"record":{"id":"rec_read","conversation_id":"conv_read","kind":"message.user"},"content":{"text":"hello"}}`))
		case "/v1/history/search":
			if request.URL.Query().Get("q") != "hello world" || request.URL.Query().Get("tag") != "env=test" {
				t.Errorf("search query = %s", request.URL.RawQuery)
			}
			_, _ = writer.Write([]byte(`{"hits":[{"record":{"id":"rec_read","conversation_id":"conv_read","kind":"message.user"},"score":1.5,"snippet":"hello world"}],"total":1,"took_ms":0.2,"indexed_records":1,"index_updated_at":"2026-01-01T00:00:00Z"}`))
		case "/v1/history/conversations":
			if request.URL.Query().Get("limit") != "10" {
				t.Errorf("query = %s", request.URL.RawQuery)
			}
			_, _ = writer.Write([]byte(`[{"id":"conv_read","record_count":1,"first_seen_at":"2026-01-01T00:00:00Z","last_seen_at":"2026-01-01T00:00:00Z","agents":[],"kinds":["message.user"]}]`))
		case "/v1/history/conversations/conv_read/timeline":
			_, _ = writer.Write([]byte(`[{"record":{"id":"rec_read","conversation_id":"conv_read","kind":"message.user"},"content":{"text":"hello"}}]`))
		default:
			http.NotFound(writer, request)
		}
	}))
	defer server.Close()

	client, err := New(WithEndpoint(server.URL), WithRetries(0, 0))
	if err != nil {
		t.Fatal(err)
	}
	ctx := context.Background()
	if err := client.Health(ctx); err != nil {
		t.Fatal(err)
	}
	records, err := client.Query(ctx, HistoryQuery{ConversationID: "conv_read", Tags: map[string]string{"env": "test"}, Limit: 25})
	if err != nil || len(records) != 1 || records[0].ID != "rec_read" {
		t.Fatalf("records = %#v, err = %v", records, err)
	}
	search, err := client.Search(ctx, SearchQuery{Text: "hello world", Tags: map[string]string{"env": "test"}, Limit: 25})
	if err != nil || search.Total != 1 || search.Hits[0].Record.ID != "rec_read" {
		t.Fatalf("search = %#v, err = %v", search, err)
	}
	entry, err := client.GetRecord(ctx, "rec_read")
	if err != nil || entry.Record.ID != "rec_read" || string(entry.Content) != `{"text":"hello"}` {
		t.Fatalf("entry = %#v, err = %v", entry, err)
	}
	conversations, err := client.Conversations(ctx, 10)
	if err != nil || len(conversations) != 1 || conversations[0].ID != "conv_read" {
		t.Fatalf("conversations = %#v, err = %v", conversations, err)
	}
	timeline, err := client.Timeline(ctx, "conv_read")
	if err != nil || len(timeline) != 1 || timeline[0].Record.ID != "rec_read" {
		t.Fatalf("timeline = %#v, err = %v", timeline, err)
	}
}

func TestAPIErrorIsTyped(t *testing.T) {
	t.Parallel()
	server := httptest.NewServer(http.HandlerFunc(func(writer http.ResponseWriter, _ *http.Request) {
		writer.WriteHeader(http.StatusConflict)
		_, _ = writer.Write([]byte(`{"error":{"code":"FH_IDEMPOTENCY_CONFLICT","message":"reused","retryable":false}}`))
	}))
	defer server.Close()
	client, _ := New(WithEndpoint(server.URL), WithRetries(0, time.Millisecond))
	_, err := client.Capture(context.Background(), CaptureInput{ConversationID: "conv", Kind: "event", Content: nil})
	var apiError *APIError
	if !errors.As(err, &apiError) || apiError.StatusCode != http.StatusConflict || apiError.Code != "FH_IDEMPOTENCY_CONFLICT" || apiError.Retryable {
		t.Fatalf("error = %#v", err)
	}
}

func TestRetryAfterZeroIsExplicit(t *testing.T) {
	delay, explicit := retryAfterDuration("0", time.Now())
	if !explicit || delay != 0 {
		t.Fatalf("delay = %s, explicit = %v", delay, explicit)
	}
}

File: sdk/typescript/test/client.test.ts
import assert from "node:assert/strict";
import { createServer, type IncomingMessage, type ServerResponse } from "node:http";
import { afterEach, beforeEach, test } from "node:test";
import {
  BatchTraceProcessor,
  setTraceProcessors,
  withAgentSpan,
  withFunctionSpan,
  withGenerationSpan,
  withTrace,
} from "@openai/agents";
import type {
  HookCallbackMatcher,
  PostToolUseHookInput,
  PreToolUseHookInput,
  UserPromptSubmitHookInput,
} from "@anthropic-ai/claude-agent-sdk";

import {
  APIError,
  BackgroundProcessor,
  DroppedEvent,
  Farfield,
  FarfieldClaudeAgentHooks,
  FarfieldOpenAIAgentsExporter,
  type Event,
  type Json,
  type WireEvent,
} from "../src/index.js";

interface Received {
  method: string;
  path: string;
  body: string;
  authorization?: string;
}

let endpoint = "";
let closeServer: (() => Promise<void>) | undefined;
let requests: Received[] = [];
let recordFailures = 0;

beforeEach(async () => {
  requests = [];
  recordFailures = 0;
  const server = createServer(async (request, response) => handle(request, response));
  await new Promise<void>((resolve) => server.listen(0, "127.0.0.1", resolve));
  const address = server.address();
  if (!address || typeof address === "string") throw new Error("expected TCP address");
  endpoint = `http://127.0.0.1:${address.port}`;
  closeServer = () => new Promise((resolve, reject) => server.close((error) => (error ? reject(error) : resolve())));
});

afterEach(async () => {
  setTraceProcessors([]);
  await closeServer?.();
});

test("capture scopes metadata and retries the identical body", async () => {
  recordFailures = 1;
  const ff = new Farfield({
    endpoint,
    token: "secret",
    retryDelayMs: 0,
    defaults: { agent: "planner", tags: { env: "test" } },
  });

  const record = await ff.withConversation(
    { id: "conv_one", traceId: "trace_one", tags: { tenant: "acme" } },
    () => ff.capture({ kind: "message.user", content: { text: "hello" } }),
  );

  assert.match(record.id, /^rec_/);
  assert.equal(requests.length, 2);
  assert.equal(requests[0]?.body, requests[1]?.body);
  assert.equal(requests[0]?.authorization, "Bearer secret");
  const sent = JSON.parse(requests[0]!.body) as WireEvent;
  assert.equal(sent.conversation_id, "conv_one");
  assert.equal(sent.trace_id, "trace_one");
  assert.equal(sent.agent, "planner");
  assert.deepEqual(sent.tags, { env: "test", tenant: "acme" });
});

test("async conversation scopes are isolated", async () => {
  const ff = new Farfield({ endpoint });
  await Promise.all(
    ["conv_a", "conv_b"].map((conversationId) =>
      ff.withConversation(conversationId, async () => {
        await new Promise((resolve) => setImmediate(resolve));
        await ff.capture({ kind: "message.user", content: conversationId });
      }),
    ),
  );

  const ids = requests.map((request) => (JSON.parse(request.body) as WireEvent).conversation_id).sort();
  assert.deepEqual(ids, ["conv_a", "conv_b"]);
});

test("conversation batches form one durable segment", async () => {
  const ff = new Farfield({ endpoint });
  const segment = await ff.conversation("conv_batch").batch(async (batch) => {
    batch.message("user", "hello");
    await new Promise((resolve) => setImmediate(resolve));
    batch.toolResult("weather", { sunny: true });
  }, "seg_one");

  assert.equal(segment.id, "seg_one");
  assert.equal(segment.entries.length, 2);
  const sent = JSON.parse(requests[0]!.body) as { records: WireEvent[] };
  assert.equal(sent.records[0]?.conversation_id, "conv_batch");
  assert.equal(sent.records[1]?.tool, "weather");
});

test("beforeSend can redact or drop before transport", async () => {
  const ff = new Farfield({
    endpoint,
    beforeSend(event) {
      if (event.kind === "debug") return null;
      return { ...event, content: "[redacted]" };
    },
  });

  await ff.capture({ conversationId: "conv_one", kind: "message.user", content: "secret" });
  await assert.rejects(
    ff.capture({ conversationId: "conv_one", kind: "debug", content: "noise" }),
    DroppedEvent,
  );
  assert.equal((JSON.parse(requests[0]!.body) as WireEvent).content, "[redacted]");
  assert.equal(requests.length, 1);
});

test("invalid JSON is rejected before transport", async () => {
  const ff = new Farfield({ endpoint });
  await assert.rejects(
    ff.capture({ conversationId: "conv_one", kind: "metric", content: Number.NaN as unknown as Json }),
    TypeError,
  );
  assert.equal(requests.length, 0);
});

test("background processor snapshots scope, batches by conversation, and flushes", async () => {
  const ff = new Farfield({ endpoint, defaults: { tags: { env: "test" } } });
  const processor = new BackgroundProcessor(ff, { maxBatchSize: 10, scheduleDelayMs: 5 });
  await ff.withConversation({ id: "conv_one", agent: "researcher" }, async () => {
    assert.equal(await processor.submit({ kind: "message.user", content: "hello" }), true);
    assert.equal(await processor.submit({ kind: "message.assistant", content: "hi" }), true);
  });
  assert.equal(
    await processor.submit({ conversationId: "conv_two", kind: "tool.result", tool: "search", content: { ok: true } }),
    true,
  );
  assert.equal(await processor.flush(), true);
  assert.equal(await processor.shutdown(), true);

  const segments = requests
    .filter((request) => request.path === "/v1/history/segments")
    .map((request) => JSON.parse(request.body) as { records: WireEvent[] });
  assert.equal(segments.length, 2);
  const groups = new Map(segments.map((segment) => [segment.records[0]!.conversation_id, segment.records]));
  assert.equal(groups.get("conv_one")?.length, 2);
  assert.equal(groups.get("conv_one")?.[0]?.agent, "researcher");
  assert.deepEqual(groups.get("conv_one")?.[0]?.tags, { env: "test" });
  assert.deepEqual(processor.stats(), {
    enqueued: 3,
    committed: 3,
    dropped: 0,
    failed: 0,
    batches: 2,
    pending: 0,
  });
});

test("background processor reports delivery failures", async () => {
  const errors: unknown[] = [];
  const ff = new Farfield({ endpoint: "http://127.0.0.1:1", maxRetries: 0, timeoutMs: 50 });
  const processor = new BackgroundProcessor(ff, {
    scheduleDelayMs: 0,
    onError(error) {
      errors.push(error);
    },
  });
  assert.equal(await processor.submit({ conversationId: "conv_fail", kind: "message.user", content: "hello" }), true);
  assert.equal(await processor.flush(), false);
  assert.equal(await processor.shutdown(), false);
  assert.equal(processor.stats().failed, 1);
  assert.equal(errors.length, 1);
});

test("OpenAI Agents SDK trace lifecycle is captured through its real exporter API", async () => {
  const exporter = new FarfieldOpenAIAgentsExporter(new Farfield({ endpoint }), { defaultAgent: "demo" });
  const processor = new BatchTraceProcessor(exporter, { scheduleDelay: 60_000 });
  setTraceProcessors([processor]);

  await withTrace(
    "research workflow",
    async () =>
      withAgentSpan(
        async () => {
          await withGenerationSpan(async () => undefined, {
            spanId: "span_1123456789abcdef0123456789abcdef",
            data: {
              input: [{ role: "user", content: "Find an answer" }],
              output: [{ role: "assistant", content: "I found it" }],
              model: "gpt-test",
              usage: { input_tokens: 4, output_tokens: 3 },
            },
          });
          await withFunctionSpan(async () => undefined, {
            spanId: "span_2123456789abcdef0123456789abcdef",
            data: { name: "web_search", input: '{"q":"Farfield"}', output: '{"results":1}' },
          });
        },
        { spanId: "span_0123456789abcdef0123456789abcdef", data: { name: "researcher" } },
      ),
    {
      traceId: "trace_0123456789abcdef0123456789abcdef",
      groupId: "conversation_openai_agents",
      metadata: { tenant: "test" },
    },
  );
  await processor.forceFlush();
  await processor.shutdown();
  await exporter.shutdown();

  const segments = requests
    .filter((request) => request.path === "/v1/history/segments")
    .map((request) => JSON.parse(request.body) as { records: WireEvent[] });
  const records = segments.flatMap((segment) => segment.records);
  assert.equal(records.length, 4);
  assert.deepEqual(new Set(records.map((record) => record.conversation_id)), new Set(["conversation_openai_agents"]));
  assert.deepEqual(
    new Set(records.map((record) => record.kind)),
    new Set(["agent.trace", "agent.invoke", "model.generation", "tool.execution"]),
  );
  assert.equal(records.find((record) => record.kind === "tool.execution")?.tool, "web_search");
  assert.deepEqual(exporter.stats(), {
    traces: 1,
    spans: 3,
    bufferedSpans: 0,
    cachedTraces: 0,
    failedExports: 0,
  });
});

test("Claude Agent SDK hook lifecycle is captured through its real callback types", async () => {
  const integration = new FarfieldClaudeAgentHooks(new Farfield({ endpoint }), {
    defaultAgent: "claude-code",
    scheduleDelayMs: 5,
  });
  const matchers = integration.matchers({ timeout: 2 });
  const preMatcher: HookCallbackMatcher = matchers.PreToolUse![0]!;
  const context = { signal: new AbortController().signal };
  const common = {
    session_id: "session_claude_agent",
    transcript_path: "/tmp/transcript.jsonl",
    cwd: "/workspace",
  };
  const prompt = {
    ...common,
    hook_event_name: "UserPromptSubmit",
    prompt: "Inspect the repo",
  } satisfies UserPromptSubmitHookInput;
  const pre = {
    ...common,
    hook_event_name: "PreToolUse",
    tool_name: "Read",
    tool_input: { file_path: "README.md" },
    tool_use_id: "toolu_01",
  } satisfies PreToolUseHookInput;
  const post = {
    ...common,
    hook_event_name: "PostToolUse",
    tool_name: "Read",
    tool_input: { file_path: "README.md" },
    tool_response: { content: "Farfield" },
    tool_use_id: "toolu_01",
  } satisfies PostToolUseHookInput;

  await matchers.UserPromptSubmit![0]!.hooks[0]!(prompt, undefined, context);
  await preMatcher.hooks[0]!(pre, "toolu_01", context);
  await matchers.PostToolUse![0]!.hooks[0]!(post, "toolu_01", context);
  assert.equal(await integration.shutdown(), true);

  const records = requests
    .filter((request) => request.path === "/v1/history/segments")
    .flatMap((request) => (JSON.parse(request.body) as { records: WireEvent[] }).records);
  assert.equal(records.length, 3);
  assert.deepEqual(new Set(records.map((record) => record.conversation_id)), new Set(["session_claude_agent"]));
  assert.deepEqual(new Set(records.map((record) => record.kind)), new Set(["message.user", "tool.call", "tool.result"]));
  const toolResult = records.find((record) => record.kind === "tool.result");
  assert.equal(toolResult?.tool, "Read");
  assert.equal((toolResult?.content as { schema: string }).schema, "farfield.claude_agent_sdk.hook.v1");
  assert.equal(integration.stats().committed, 3);
});

test("history reads cover the complete API", async () => {
  const ff = new Farfield({ endpoint });
  assert.equal(await ff.health(), true);
  assert.equal((await ff.query({ conversationId: "conv_query", tags: { env: "test" }, limit: 10 }))[0]?.id, "rec_query");
  assert.equal((await ff.search({ text: "hello", tags: { env: "test" } })).hits[0]?.record.id, "rec_query");
  assert.deepEqual((await ff.getRecord("rec_query")).content, { text: "hello" });
  assert.equal((await ff.conversations())[0]?.id, "conv_query");
  assert.equal((await ff.timeline("conv_query"))[0]?.record.id, "rec_query");

});

test("server failures are typed and actionable", async () => {
  const ff = new Farfield({ endpoint, maxRetries: 0 });
  await assert.rejects(
    ff.capture({ conversationId: "error", kind: "message.user", content: null }),
    (error) => error instanceof APIError && error.statusCode === 409 && error.code === "FH_CONFLICT",
  );
});

async function handle(request: IncomingMessage, response: ServerResponse): Promise<void> {
  const body = await readBody(request);
  requests.push({
    method: request.method ?? "",
    path: request.url ?? "",
    body,
    ...(request.headers.authorization ? { authorization: request.headers.authorization } : {}),
  });
  const url = new URL(request.url ?? "/", "http://localhost");
  if (request.method === "POST" && url.pathname === "/v1/history/records") {
    const value = JSON.parse(body) as WireEvent;
    if (value.conversation_id === "error") return sendError(response, 409, "FH_CONFLICT", "already exists");
    if (recordFailures > 0) {
      recordFailures -= 1;
      return sendError(response, 503, "FH_BUSY", "try again", true);
    }
    return send(response, 201, record(value));
  }
  if (request.method === "POST" && url.pathname === "/v1/history/segments") {
    const value = JSON.parse(body) as { id: string; records: WireEvent[] };
    return send(response, 201, {
      schema_version: "farfield.history.segment.v1",
      id: value.id,
      conversation_id: value.records[0]?.conversation_id,
      entries: value.records.map((event) => ({ record: record(event), content: event.content })),
      segment_sha256: "b".repeat(64),
    });
  }
  if (url.pathname === "/v1/health") return send(response, 200, { ok: true, service: "farfield" });
  if (url.pathname === "/v1/history/records" && request.method === "GET") {
    assert.equal(url.searchParams.get("tag"), "env=test");
    return send(response, 200, [record({ id: "rec_query", conversation_id: "conv_query", kind: "message.user", content: null } as WireEvent)]);
  }
  if (url.pathname === "/v1/history/search") {
    assert.equal(url.searchParams.get("q"), "hello");
    assert.equal(url.searchParams.get("tag"), "env=test");
    return send(response, 200, { hits: [{ record: record({ id: "rec_query", conversation_id: "conv_query", kind: "message.user", content: null } as WireEvent), score: 1, snippet: "hello" }], total: 1, took_ms: 0.2, indexed_records: 1, index_updated_at: "2026-01-01T00:00:00Z" });
  }
  if (url.pathname === "/v1/history/records/rec_query") return send(response, 200, { record: record({ id: "rec_query", conversation_id: "conv_query", kind: "message.user", content: null } as WireEvent), content: { text: "hello" } });
  if (url.pathname === "/v1/history/conversations") return send(response, 200, [{ id: "conv_query", record_count: 1, first_seen_at: "2026-01-01T00:00:00Z", last_seen_at: "2026-01-01T00:00:00Z", agents: [], kinds: ["message.user"] }]);
  if (url.pathname === "/v1/history/conversations/conv_query/timeline") return send(response, 200, [{ record: record({ id: "rec_query", conversation_id: "conv_query", kind: "message.user", content: null } as WireEvent), content: { text: "hello" } }]);
  sendError(response, 404, "FH_NOT_FOUND", "not found");
}

function record(event: WireEvent): Record<string, unknown> {
  return {
    ...event,
    schema_version: "farfield.history.record.v2",
    occurred_at: "2026-01-01T00:00:00Z",
    recorded_at: "2026-01-01T00:00:00Z",
    record_sha256: "a".repeat(64),
    tags: {},
  };
}

function readBody(request: IncomingMessage): Promise<string> {
  return new Promise((resolve, reject) => {
    const chunks: Buffer[] = [];
    request.on("data", (chunk: Buffer) => chunks.push(chunk));
    request.on("end", () => resolve(Buffer.concat(chunks).toString()));
    request.on("error", reject);
  });
}

function send(response: ServerResponse, status: number, value: unknown): void {
  const body = JSON.stringify(value);
  response.writeHead(status, { "Content-Type": "application/json", "Content-Length": Buffer.byteLength(body) });
  response.end(body);
}

function sendError(response: ServerResponse, status: number, code: string, message: string, retryable = false): void {
  send(response, status, { error: { code, message, retryable } });
}
